build.sbt

```
@@ -184,7 +188,20 @@ lazy val commonSettings = List(
   resolvers ++= Resolver.sonatypeOssRepos("snapshots"),
   startYear := Some(2019),
   apiURL := Some(url("https://typelevel.org/cats-tagless/api/")),
-  autoAPIMappings := true
+  autoAPIMappings := true,
+  scalacOptions ++= (CrossVersion.partialVersion(scalaVersion.value) match {
+    case Some((3, _)) =>
+      Seq(
+        "-language:adhocExtensions",
+        "-source:future",
+        "-explain"
+      )
+    case Some((2, 12 | 13)) => Seq("-Xsource:3", "-P:kind-projector:underscore-placeholders")
+    case _ => Nil
+  }),
+  // sbt-typelevel sets -source:3.0-migration, we'd like to replace it with -source:future
+  scalacOptions := scalacOptions.value.filterNot(_ == "-source:3.0-migration"),
```

I have since learned about it:
```suggestion
  scalacOptions ~= (_.filterNot(_ == "-source:3.0-migration")),
```